apps/api/src/domain/models/song/commands/publish-song.command-handler.ts

```diff
@@ -0,0 +1,85 @@
+import { Ack, CommandHandler, ICommandHandler } from '@coscrad/commands';
+import { buildSimpleValidationFunction } from '@coscrad/validation';
+import { Inject } from '@nestjs/common';
+import { InternalError, isInternalError } from '../../../../lib/errors/InternalError';
+import { isNotFound, NotFound } from '../../../../lib/types/not-found';
+import { RepositoryProvider } from '../../../../persistence/repositories/repository.provider';
+import getInstanceFactoryForEntity from '../../../factories/getInstanceFactoryForEntity';
+import { IIdGenerator } from '../../../interfaces/id-generator.interface';
+import { ResourceType } from '../../../types/ResourceType';
+import InvalidCommandPayloadTypeError from '../../shared/common-command-errors/InvalidCommandPayloadTypeError';
+import { Song } from '../song.entity';
+import { PublishSong } from './publish-song.command';
+import { SongPublished } from './song-published.event';
+
+@CommandHandler(PublishSong)
+export class PublishSongCommandHandler implements ICommandHandler {
+    constructor(
+        private readonly repositoryProvider: RepositoryProvider,
+        @Inject('ID_GENERATOR') private readonly idGenerator: IIdGenerator
+    ) {}
+
+    async execute(command: PublishSong): Promise<Ack | InternalError> {
+        const payloadTypeErrors = buildSimpleValidationFunction(
+            Object.getPrototypeOf(command).constructor
+        )(command).map(
+            (simpleError) => new InternalError(`invalid payload type: ${simpleError.toString()}`)
+        );
+
+        if (payloadTypeErrors.length > 0) {
+            return new InvalidCommandPayloadTypeError('PUBLISH_SONG', payloadTypeErrors);
+        }
+
+        // validate external state
+        const { id } = command;
+
+        const allSongs = await this.repositoryProvider
+            .forResource<Song>(ResourceType.song)
+            .fetchMany();
+
+        // TODO We need a way to include `unpublished` results in `fetchById`
+        const searchResult =
+            allSongs
```

Comment:
I'm sure this has to do with ddd, but why would you return a list of all songs and then filter them to find the one with the id from the payload rather than doing a fetchbyid to try to return just the song that you want to publish?

Reply:
Also note that whatever we do for publication status could also apply to soft deletes later.